src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala

```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+    def visitAnnotationss(annotss: Array[java.util.List[AnnotationNode]]) = if (annotss != null) annotss foreach visitAnnotations
+
+    def visitHandle(handle: Handle): Unit = {
+      visitInternalNameOrArrayReference(handle.getOwner)
+      visitDescriptor(handle.getDesc)
+    }
+  }
+
+  abstract class GenericSignatureVisitor {
+    def visitInternalName(internalName: InternalName): Unit
+
+    def raiseError(msg: String, sig: String, e: Option[Throwable] = None): Unit
+
+    def visitClassSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.classSignature() }
+    }
+
+    def visitMethodSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.methodSignature() }
+    }
+
+    def visitFieldSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.fieldSignature() }
+    }
+
+    private final class Parser(sig: String) {
+      private var index = 0
+      private val end = sig.length
+
+      private val Aborted: Throwable = new NoStackTrace { }
+      private def abort(): Nothing = throw Aborted
+
+      def safely(f: => Unit): Unit = try f catch {
+        case Aborted =>
+        case NonFatal(e) => raiseError(s"Exception thrown during signature parsing", sig, Some(e))
+      }
+
+      private def current = {
+        if (index >= end) {
+          raiseError(s"Out of bounds, $index >= $end", sig)
+          abort() // Don't continue, even if `notifyInvalidSignature` returns
+        }
+        sig.charAt(index)
+      }
+
+      private def accept(c: Char): Unit = {
+        if (current != c) {
+          raiseError(s"Expected $c at $index, found $current", sig)
+          abort()
+        }
+        index += 1
+      }
+
+      private def skip(): Unit = { index += 1 }
+      private def getCurrentAndSkip(): Char = { val c = current; skip(); c }
+
+      private def skipUntil(isDelimiter: Char => Boolean): Unit = {
+        while (!isDelimiter(current)) { index += 1 }
+      }
+
+      private def appendUntil(builder: java.lang.StringBuilder, isDelimiter: Char => Boolean): Unit = {
+        val start = index
+        skipUntil(isDelimiter)
+        builder.append(sig, start, index)
+      }
+
+      def isBaseType(c: Char): Boolean = c match {
+        case 'B' | 'C' | 'D' | 'F' | 'I' | 'J' | 'S' | 'Z' => true
+        case _ => false
+      }
+
+      private val isClassNameEnd = (c: Char) => c == '<' || c == '.' || c == ';'
+
+      private def typeArguments(): Unit = if (current == '<') {
+        skip()
+        while (current != '>') current match {
+          case '*' | '+' | '-' =>
+            skip()
+          case _ =>
+            referenceTypeSignature()
+        }
+        accept('>')
+      }
+
+      @tailrec private def referenceTypeSignature(): Unit = getCurrentAndSkip() match {
+        case 'L' =>
+          val names = new java.lang.StringBuilder()
```

Comment:
Based on a [back of the envelope](https://docs.google.com/spreadsheets/d/1uzsap5f7UV-Z42mivUqk1Ly38cE5VvDGh4geMgsTzq0/edit?usp=sharing) analysis of SCALA_HOME/lib and a selection of JARs from my ivy cache, I'd say 64. would be preferable.